qiskit_nature/operators/second_quantization/sparse_label_op.py

```diff
@@ -0,0 +1,137 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The Sparse Label Operator base class."""
+
+from __future__ import annotations
+from typing import Mapping, Iterator
+from numbers import Number
+import cmath
+import numpy as np
+
+from qiskit.quantum_info.operators.mixins import LinearMixin, AdjointMixin, TolerancesMixin
+
+
+class SparseLabelOp(LinearMixin, AdjointMixin, TolerancesMixin):
+    """The Sparse Label Operator base class."""
+
+    def __init__(self, data: Mapping[str, Number], register_length: int):
+        self._data = data
+        self._register_length = register_length
+
+    def _add(self, other: SparseLabelOp, qargs=None) -> SparseLabelOp:
+        """Return Operator addition of self and other.
+
+        Args:
+            other: the second ``SparseLabelOp`` to add to the first.
+
+        Returns:
+            the new summed ``SparseLabelOp``.
+
+        Raises:
+            ValueError: when ``qargs`` argument is not ``None``
+        """
+        if qargs is not None:
```

Comment:
If other is not a SparseLabelOp instance we should raise a ValueError (Python seeing A + B will effectively try A.add(B) if that raises a TypeError will try B.add(A) since B may support adding an A even if A can't add a B). This then analogous to what is done in multiply where other is type checked against what is supported.